network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -196,6 +197,57 @@ func (c *ControlMsgValidationInspector) Inspect(from peer.ID, rpc *pubsub.RPC) e
 	return nil
 }
 
+// inspectIWant inspects RPC iWant control messages. This func will sample the iWants and perform validation on each iWant in the sample.
+// Ensuring that the following are true:
+// - Each iWant corresponds to an iHave that was sent.
+// - Each topic in the iWant sample is a valid topic.
+// If the number of iWants that do not have a corresponding iHave exceed the configured threshold an error is returned.
+// Args:
+// - iWant: the list of iWant control messages.
+// Returns:
+// - DuplicateFoundErr: if there are any duplicate message ids found in across any of the iWants.
+// - IWantCacheMissThresholdErr: if the rate of cache misses exceeds the configured allowed threshold.
+// - error: if any error occurs while sampling the iWants
+func (c *ControlMsgValidationInspector) inspectIWant(iWants []*pubsub_pb.ControlIWant) error {
+	sampleSize := uint(10 * c.rpcTracker.LastHighestIHaveRPCSize())
+	if sampleSize == 0 || sampleSize > c.config.IWantRPCInspectionConfig.MaxSampleSize {
+		sampleSize = c.config.IWantRPCInspectionConfig.MaxSampleSize
+	}
+
+	swap := func(i, j uint) {
+		iWants[i], iWants[j] = iWants[j], iWants[i]
+	}
```

Comment:
⚠️ Please note that an RPC can carry an array of `iWant` messages, with each `iWant` potentially containing numerous message IDs. When sampling `iWant`s from an RPC, we should consider all message IDs encompassed within the `iWants []*pubsub_pb.ControlIWant` as a collective pool from which we sample. Conceptualize an `iWant` as a box filled with many message IDs, making `iWants []*pubsub_pb.ControlIWant` a collection of such boxes. The current method samples `sampleSize` boxes and subsequently checks their entire contents. This approach has three notable drawbacks:

1. Typically, `10 * c.rpcTracker.LastHighestIHaveRPCSize()` significantly surpasses the number of `iWant` boxes. Consequently, this method inspects the _entire contents of all boxes_, which is not the desired outcome and can be exploited in spam attacks.
2. Each box is examined in isolation, disregarding potential duplicate message IDs that might exist across different boxes or multiple instances of the same box.
3. The code iterates over all contents of the sampled box, which can lead to a resource exhaustion attack if a box contains a large but finite number of message IDs.


Suggestion, open all boxes and drop all message ids in an `iWant` pool and then sample from that `iWant` pool: 
```go
	var iWantPool []*pubsub_pb.ControlIWant
	// opens all iWant boxes into a sample pool to be sampled.
	for _, iWant := range iWants {
		if len(iWant.GetMessageIDs()) == 0 {
			continue
		}
		iWantPool = append(iWantPool, iWant)
	}

	swap := func(i, j uint) {
		iWantPool[i], iWantPool[j] = iWantPool[j], iWantPool[i]
	}
```

Reply:
https://github.com/onflow/flow-go/pull/4574/commits/3e998426a232997e1eb3933f6b036962f02e448e